File: dashboard/routes/shares.py
"""Links that reach one machine, and the guest page behind them."""

from flask import Blueprint, Response, abort, g, jsonify, render_template, request

import audit
import fleet
import shares
from db import connect
from machines import computer_view, get_computer, list_computers
from screens import bridge_screenshot
from security import require_token
from settings import SHOT_TTL


bp = Blueprint("shares", __name__)


def share_base_url() -> str:
    return request.host_url.rstrip("/")


def share_view(row: dict, comp_name: str | None = None) -> dict:
    out = dict(row)
    out.pop("token_hash", None)
    out["status"] = shares.status(row)
    out["url"] = f"{share_base_url()}/s/{row['token']}"
    if comp_name:
        out["computer"] = comp_name
    return out


@bp.route("/api/v1/shares", methods=["GET"])
def api_shares_list():
    conn = connect()
    names = {c["id"]: c["name"] for c in list_computers()}
    rows = [share_view(r, names.get(r["computer_id"])) for r in shares.listing(conn)]
    conn.close()
    return jsonify({"ok": True, "data": rows, "error": None})


@bp.route("/api/v1/computers/<int:comp_id>/shares", methods=["POST"])
@require_token
def api_share_create(comp_id: int):
    comp = get_computer(comp_id)
    if not comp:
        return jsonify({"ok": False, "data": None, "error": "not found"}), 404
    payload = request.get_json(silent=True) or {}
    conn = connect()
    try:
        row = shares.create(conn, comp_id,
                            label=payload.get("label", ""),
                            mode=(payload.get("mode") or "watch").strip(),
                            hours=payload.get("hours"))
    except ValueError as exc:
        conn.close()
        return jsonify({"ok": False, "data": None, "error": str(exc)}), 400
    conn.close()
    g.audit_detail = f"{row['mode']} share '{row['label']}' until {row['expires_at']}"
    return jsonify({"ok": True, "data": share_view(row, comp["name"]), "error": None}), 201


@bp.route("/api/v1/shares/<int:share_id>", methods=["DELETE"])
@require_token
def api_share_revoke(share_id: int):
    conn = connect()
    row = conn.execute("SELECT * FROM shares WHERE id = ?", (share_id,)).fetchone()
    if not row:
        conn.close()
        return jsonify({"ok": False, "data": None, "error": "not found"}), 404
    revoked = shares.revoke(conn, share_id)
    conn.close()
    g.audit_detail = f"share '{row['label']}' ({row['mode']})"
    return jsonify({"ok": True, "error": None, "data": {
        "id": share_id, "revoked": revoked,
        # Being straight about what revoking a control share does and doesn't
        # do matters more than sounding reassuring.
        "note": ("the guest's browser already holds this machine's screen "
                 "password — rotate it to retract access completely")
        if row["mode"] == "control" else None,
    }})


@bp.route("/api/v1/computers/<int:comp_id>/rotate-password", methods=["POST"])
@require_token
def api_rotate_password(comp_id: int):
    """Give the machine a new screen password and restart it.

    This is the hard revoke behind a control share: anyone holding the old
    noVNC URL is out, including a guest who saved it.
    """
    comp = get_computer(comp_id)
    if not comp:
        return jsonify({"ok": False, "data": None, "error": "not found"}), 404
    password = fleet.random_vnc_password()
    try:
        fleet.destroy_computer(comp["slug"], keep_home=True)
        fleet.create_computer(comp["slug"], comp["novnc_port"], password, image=comp["image"])
    except Exception as exc:  # noqa: BLE001
        return jsonify({"ok": False, "data": None,
                        "error": f"failed to rotate: {exc}"}), 500
    conn = connect()
    conn.execute("UPDATE computers SET vnc_password = ? WHERE id = ?", (password, comp_id))
    conn.execute("UPDATE shares SET revoked = 1 WHERE computer_id = ? AND mode = 'control'",
                 (comp_id,))
    conn.commit()
    conn.close()
    return jsonify({"ok": True, "data": {"id": comp_id, "rotated": True}, "error": None})


def current_share():
    """The live share behind this request, or None."""
    token = request.view_args.get("token") if request.view_args else None
    conn = connect()
    row = shares.resolve(conn, token or "")
    if row:
        shares.note_use(conn, row["id"], request.remote_addr)
        g.actor = f"share:{row['label']}"
    conn.close()
    return row


@bp.route("/s/<token>")
def share_page(token: str):
    """The page a guest sees. One machine, nothing else, no dashboard."""
    row = current_share()
    if not row:
        return render_template("share_gone.html"), 404
    comp = get_computer(row["computer_id"])
    if not comp:
        return render_template("share_gone.html"), 404
    view = computer_view(comp)
    audit.record(f"GET /s/<token> ({row['mode']})", actor=f"share:{row['label']}",
                 source_ip=request.remote_addr, target=comp["name"],
                 detail="opened the share page", status=200)
    return render_template("share.html", comp=comp, view=view, share=row, token=token)


@bp.route("/s/<token>/screen.png")
def share_screen(token: str):
    """The screen, served through the share rather than the machine's port —
    which is what makes a `watch` share fully revocable."""
    row = current_share()
    if not row:
        abort(404)
    comp = get_computer(row["computer_id"])
    if not comp:
        abort(404)
    if not fleet.is_running(comp["slug"]):
        abort(503)
    png = bridge_screenshot(computer_view(comp, with_state=False))
    if png is None:
        abort(503)
    return Response(png, mimetype="image/png",
                    headers={"Cache-Control": f"max-age={int(SHOT_TTL)}"})

File: dashboard-ts/src/routes/shares.ts
/** Links that reach one machine, and the guest page behind them. */

import { Hono, type Context } from "hono";
import * as audit from "./../audit";
import * as fleet from "./../fleet";
import * as sharesLib from "./../shares";
import { one, run } from "./../db";
import {
  computerView,
  getComputer,
  listComputers,
  type Computer,
} from "./../machines";
import { bridgeScreenshot } from "./../screens";
import { render } from "./../templates";
import { requireToken } from "./../security";
import { SHOT_TTL } from "./../settings";
import {
  browserHost,
  fail,
  intParam,
  jsonBody,
  notFound,
  ok,
  sourceIp,
  type Env,
} from "./../http";

export const shares = new Hono<Env>();

function loadComputer(c: Context<Env>): Computer | null {
  const id = intParam(c, "id");
  if (id === null) return null;
  const comp = getComputer(id);
  if (comp) c.set("auditTarget", comp.name);
  return comp;
}

function shareBaseUrl(c: Context<Env>): string {
  const url = new URL(c.req.url);
  return `${url.protocol}//${c.req.header("host") ?? url.host}`;
}

function shareView(c: Context<Env>, row: sharesLib.ShareRow, compName?: string | null) {
  const { token_hash, ...rest } = row;
  return {
    ...rest,
    status: sharesLib.status(row),
    url: `${shareBaseUrl(c)}/s/${row.token}`,
    ...(compName ? { computer: compName } : {}),
  };
}

shares.get("/api/v1/shares", (c) => {
  const names = new Map(listComputers().map((x) => [x.id, x.name]));
  return ok(
    c,
    sharesLib.listing().map((r) => shareView(c, r, names.get(r.computer_id))),
  );
});

shares.post("/api/v1/computers/:id/shares", requireToken, async (c) => {
  const comp = loadComputer(c);
  if (!comp) return notFound(c);
  const payload = await jsonBody(c);
  let row: sharesLib.ShareRow;
  try {
    row = sharesLib.create(
      comp.id,
      String(payload.label ?? ""),
      String(payload.mode || "watch").trim(),
      payload.hours,
    );
  } catch (err: any) {
    if (err instanceof sharesLib.ValidationError) return fail(c, String(err.message));
    throw err;
  }
  c.set("auditDetail", `${row.mode} share '${row.label}' until ${row.expires_at}`);
  return ok(c, shareView(c, row, comp.name), 201);
});

shares.delete("/api/v1/shares/:id", requireToken, (c) => {
  const id = intParam(c, "id");
  const row = id === null ? null : one<sharesLib.ShareRow>(
    "SELECT * FROM shares WHERE id = ?",
    id,
  );
  if (!row) return notFound(c);
  const revoked = sharesLib.revoke(row.id);
  c.set("auditDetail", `share '${row.label}' (${row.mode})`);
  return ok(c, {
    id: row.id,
    revoked,
    // Being straight about what revoking a control share does and doesn't do
    // matters more than sounding reassuring.
    note:
      row.mode === "control"
        ? "the guest's browser already holds this machine's screen " +
          "password — rotate it to retract access completely"
        : null,
  });
});

/**
 * Give the machine a new screen password and restart it.
 *
 * This is the hard revoke behind a control share: anyone holding the old noVNC
 * URL is out, including a guest who saved it.
 */
shares.post("/api/v1/computers/:id/rotate-password", requireToken, async (c) => {
  const comp = loadComputer(c);
  if (!comp) return notFound(c);
  const password = fleet.randomVncPassword();
  try {
    await fleet.destroyComputer(comp.slug, true);
    await fleet.createComputer(comp.slug, comp.novnc_port, password, comp.image);
  } catch (err: any) {
    return fail(c, `failed to rotate: ${err?.message ?? err}`, 500);
  }
  run("UPDATE computers SET vnc_password = ? WHERE id = ?", password, comp.id);
  run(
    "UPDATE shares SET revoked = 1 WHERE computer_id = ? AND mode = 'control'",
    comp.id,
  );
  return ok(c, { id: comp.id, rotated: true });
});

/** The live share behind this request, or null. */
function currentShare(c: Context<Env>): sharesLib.ShareRow | null {
  const row = sharesLib.resolve(c.req.param("token") ?? "");
  if (row) {
    sharesLib.noteUse(row.id, sourceIp(c) || null);
    c.set("actor", `share:${row.label}`);
  }
  return row;
}

/** The page a guest sees. One machine, nothing else, no dashboard. */
shares.get("/s/:token", async (c) => {
  const row = currentShare(c);
  if (!row) return c.html(render("share_gone.html"), 404);
  const comp = getComputer(row.computer_id);
  if (!comp) return c.html(render("share_gone.html"), 404);
  const view = await computerView(comp, { host: browserHost(c) });
  audit.record(`GET /s/<token> (${row.mode})`, {
    actor: `share:${row.label}`,
    source_ip: sourceIp(c),
    target: comp.name,
    detail: "opened the share page",
    status: 200,
  });
  return c.html(
    render("share.html", { comp, view, share: row, token: c.req.param("token") }),
  );
});

/** The screen, served through the share rather than the machine's port —
 *  which is what makes a `watch` share fully revocable. */
shares.get("/s/:token/screen.png", async (c) => {
  const row = currentShare(c);
  if (!row) return c.text("not found", 404);
  const comp = getComputer(row.computer_id);
  if (!comp) return c.text("not found", 404);
  if (!(await fleet.isRunning(comp.slug))) return c.text("unavailable", 503);
  const png = await bridgeScreenshot(await computerView(comp, { withState: false }));
  if (png === null) return c.text("unavailable", 503);
  return c.body(png as any, 200, {
    "content-type": "image/png",
    "cache-control": `max-age=${Math.trunc(SHOT_TTL)}`,
  });
});
